platform-sdk/docs/core/tipset-algorithm.md

```diff
@@ -0,0 +1,260 @@
+each entry that has a greater value, instead add the amount of consensus weight held by that node.
+
+```
+Example
+
+Suppose an address book has the following weights: A = 5, B = 9, C = 11, D = 2
+
+Find the weighed advancement score between X and Y.
+
+X = [1, 3, 5, 2]
+Y = [7, 3, 2, 11]
+
+The entry for node A advances, add 5. The entry for D advances, add 2. Total weighed advancement score is 7.
+```
+
+# Partial Weighted Tipset Advancement Score
+
+This is similar to the weighed advancement score with one minor tweak. A partial weighted advancement score
+is always calculated from the perspective of a particular node. When computing a partial advancement score,
+ignore the advancement provided by a target node.
+
+```
+Example
+
+Suppose an address book has the following weights: A = 5, B = 9, C = 11, D = 2
+
+Find the partial weighted advancement score between X and Y with a target of node A.
+
+X = [1, 3, 5, 2]
+Y = [7, 3, 2, 11]
+
+The entry for node A advances, but we ignore it because A is the target.
+The entry for D advances, add 2. Total weighed advancement score is 2.
+```
+
+# Snapshot Improvement Scores
+
+Each new event is assigned a tipset improvement score. Each node assigns the events that it creates an
+improvement score. Nodes do not assign events created by other nodes an improvement score.
+
+Keep track of a special tipset, called the "snapshot" tipset. The snapshot starts out empty, i.e.
+`[-1, -1, -1, ..., -1]` at genesis. Periodically, the snapshot is updated to a more recent tipset
+(how this happens is described below).
+
+Each time a node creates a new event, compare that event's tipset to the snapshot tipset, and find the partial
+weighted advancement score of that new event targeting the event's creator.
+
+There is a special threshold that causes the snapshot to advance. The threshold for each node is slightly different,
+unless nodes have the same consensus weight. To calculate the threshold, find the minimum weight required to have >2/3
+of the total weight, and then subtract the node's weight.
+
+```
+Example of computing snapshot advancement threshold
+
+Suppose an address book has the following weights: A = 5, B = 9, C = 11, D = 2. Compute the snapshot advancement
+threshold for node A.
+
+Total consensus weight is 27. The amount of stake required to have >2/3 of the total weight is 19. Subtract A's weight,
+and A's snapshot advancement threshold is 14.
+```
+
+When a new event has a snapshot improvement score that is greater than or equal to the snapshot advancement threshold,
+then the snapshot is updated to the tipset of that new event.
+
+# Limiting Event Creation by Snapshot Improvement Score
+
+Nodes are not permitted to create a new event unless one of the following two conditions are true:
+
+1) this is the first event a node has ever created (i.e. this is genesis), or
+2) the snapshot improvement score exceeds the snapshot improvement score of the previous event, or
```

Comment:
Wouldn't it be simpler to say, "the partial weighted tipset score is greater than zero"?

Reply:
Not quite. The partial weighted tipset score is relative to the snapshot. It's possible to have a positive advancement weight relative to the snapshot. But what's important is that each node advances more than the one before it.

Reply:
Not according to the previous paragraphs. The score computed with respect to the snapshot is a "special partial weighted advancement score" referred to as the "snapshot improvement score". The partial weighted tipset score is how much this event improves over the parent tipsets and does not specifically reference the snapshot, which is exactly what I think this is trying to communicate.